File: frontend/coordinator-android/engine/src/main/kotlin/ug/qaat/engine/RoomCode.kt
package ug.qaat.engine

import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * RoomCode — HMAC-SHA256 TOTP (RFC 6238/4226 style) port of
 * backend/api-gateway/internal/checkin/roomcode.go. The native coordinator app
 * now GENERATES the displayed code and VALIDATES the submitted one, so both must
 * match the server's algorithm exactly (10s step, 6 digits, ±1 step window).
 */
object RoomCode {
    const val STEP_SECONDS = 10L
    const val DIGITS = 6
    private const val SKEW_STEPS = 1L
    private const val MOD = 1_000_000L // 10^DIGITS

    /** The code for a per-session secret at the given epoch-seconds. */
    fun derive(secret: ByteArray, epochSeconds: Long): String =
        deriveStep(secret, epochSeconds / STEP_SECONDS)

    /** Accepts the code within ±1 step of now (constant-time compare). */
    fun validate(secret: ByteArray, code: String, epochSeconds: Long): Boolean {
        if (code.length != DIGITS) return false
        val current = epochSeconds / STEP_SECONDS
        var ok = false
        for (d in -SKEW_STEPS..SKEW_STEPS) {
            if (constantTimeEquals(deriveStep(secret, current + d), code)) ok = true
        }
        return ok
    }

    /** Seconds until the current code rotates (for the coordinator's countdown). */
    fun secondsRemaining(epochSeconds: Long): Int = (STEP_SECONDS - epochSeconds % STEP_SECONDS).toInt()

    /**
     * StaticCode — the per-session code for STUDENTS (does NOT rotate). Port of
     * checkin.StaticCode in roomcode.go: HMAC of the secret salted with a fixed label,
     * no time component. A student's proximity is proven by the mandatory hotspot LAN +
     * their device-bound QR, so their code only needs to identify the room. The lecturer
     * keeps the rotating derive()/validate().
     */
    fun staticCode(secret: ByteArray): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        val sum = mac.doFinal("student-static-v1".toByteArray(Charsets.US_ASCII))
        val offset = (sum[sum.size - 1].toInt() and 0x0f)
        val bin = ((sum[offset].toInt() and 0x7f) shl 24) or
            ((sum[offset + 1].toInt() and 0xff) shl 16) or
            ((sum[offset + 2].toInt() and 0xff) shl 8) or
            (sum[offset + 3].toInt() and 0xff)
        val code = (bin.toLong() and 0xffffffffL) % MOD
        return code.toString().padStart(DIGITS, '0')
    }

    /** Whether code equals the session's static student code (constant-time). */
    fun validateStatic(secret: ByteArray, code: String): Boolean {
        if (code.trim().length != DIGITS) return false
        return constantTimeEquals(staticCode(secret), code.trim())
    }

    private fun deriveStep(secret: ByteArray, step: Long): String {
        val msg = ByteArray(8)
        var s = step
        for (i in 7 downTo 0) { msg[i] = (s and 0xff).toByte(); s = s ushr 8 } // big-endian uint64
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(secret, "HmacSHA256"))
        val sum = mac.doFinal(msg)
        val offset = (sum[sum.size - 1].toInt() and 0x0f)
        val bin = ((sum[offset].toInt() and 0x7f) shl 24) or
            ((sum[offset + 1].toInt() and 0xff) shl 16) or
            ((sum[offset + 2].toInt() and 0xff) shl 8) or
            (sum[offset + 3].toInt() and 0xff)
        val code = (bin.toLong() and 0xffffffffL) % MOD
        return code.toString().padStart(DIGITS, '0')
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        if (a.length != b.length) return false
        var r = 0
        for (i in a.indices) r = r or (a[i].code xor b[i].code)
        return r == 0
    }
}

File: frontend/coordinator-flutter/lib/engine/room_code.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as crypto;

class RoomCode {
  static const int stepSeconds = 10;
  static const int digits = 6;
  static const int _skewSteps = 1;
  static const int _mod = 1000000;

  static String derive(List<int> secret, int epochSeconds) =>
      _deriveStep(secret, epochSeconds ~/ stepSeconds);

  static bool validate(List<int> secret, String code, int epochSeconds) {
    if (code.length != digits) return false;
    final current = epochSeconds ~/ stepSeconds;
    var ok = false;
    for (var d = -_skewSteps; d <= _skewSteps; d++) {
      if (_constantTimeEquals(_deriveStep(secret, current + d), code)) {
        ok = true;
      }
    }
    return ok;
  }

  static int secondsRemaining(int epochSeconds) =>
      stepSeconds - epochSeconds % stepSeconds;

  static String staticCode(List<int> secret) {
    final mac = crypto.Hmac(crypto.sha256, secret)
        .convert(utf8.encode('student-static-v1'))
        .bytes;
    final offset = mac[mac.length - 1] & 0x0f;
    final bin = ((mac[offset] & 0x7f) << 24) |
        ((mac[offset + 1] & 0xff) << 16) |
        ((mac[offset + 2] & 0xff) << 8) |
        (mac[offset + 3] & 0xff);
    final code = (bin & 0xffffffff) % _mod;
    return code.toString().padLeft(digits, '0');
  }

  static bool validateStatic(List<int> secret, String code) {
    if (code.trim().length != digits) return false;
    return _constantTimeEquals(staticCode(secret), code.trim());
  }

  static String _deriveStep(List<int> secret, int step) {
    final msg = Uint8List(8);
    var s = step;
    for (var i = 7; i >= 0; i--) {
      msg[i] = s & 0xff;
      s = s ~/ 256;
    }
    final mac = crypto.Hmac(crypto.sha256, secret).convert(msg).bytes;
    final offset = mac[mac.length - 1] & 0x0f;
    final bin = ((mac[offset] & 0x7f) << 24) |
        ((mac[offset + 1] & 0xff) << 16) |
        ((mac[offset + 2] & 0xff) << 8) |
        (mac[offset + 3] & 0xff);
    final code = (bin & 0xffffffff) % _mod;
    return code.toString().padLeft(digits, '0');
  }

  static bool _constantTimeEquals(String a, String b) {
    if (a.length != b.length) return false;
    var r = 0;
    for (var i = 0; i < a.length; i++) {
      r = r | (a.codeUnitAt(i) ^ b.codeUnitAt(i));
    }
    return r == 0;
  }
}